// @assistant-ui/react-ai-sdk/dist/ui/use-chat/useChatRuntime.js
"use client";

// src/ui/use-chat/useChatRuntime.tsx
import { useChat } from "@ai-sdk/react";
import {
  unstable_useCloudThreadListAdapter,
  unstable_useRemoteThreadListRuntime,
  useAssistantState
} from "@assistant-ui/react";
import {
  useAISDKRuntime
} from "./useAISDKRuntime.js";
import { AssistantChatTransport } from "./AssistantChatTransport.js";
var useChatThreadRuntime = (options) => {
  const {
    adapters,
    transport: transportOptions,
    toCreateMessage,
    ...chatOptions
  } = options ?? {};
  const transport = transportOptions ?? new AssistantChatTransport();
  const id = useAssistantState(({ threadListItem }) => threadListItem.id);
  const chat = useChat({
    ...chatOptions,
    id,
    transport
  });
  const runtime = useAISDKRuntime(chat, {
    adapters,
    ...toCreateMessage && { toCreateMessage }
  });
  if (transport instanceof AssistantChatTransport) {
    transport.setRuntime(runtime);
  }
  return runtime;
};
var useChatRuntime = ({
  cloud,
  ...options
} = {}) => {
  const cloudAdapter = unstable_useCloudThreadListAdapter({ cloud });
  return unstable_useRemoteThreadListRuntime({
    runtimeHook: function RuntimeHook() {
      return useChatThreadRuntime(options);
    },
    adapter: cloudAdapter
  });
};
export {
  useChatRuntime,
  useChatThreadRuntime
};
